### PR TITLE
FOGL-3379: Comprehensive time bucket API for graph production

### PR DESCRIPTION
**Pull request by pintomax**
FOGL-3379: Comprehensive time bucket API for graph production

test payload:

```
{
        "where" : {
                        "column" : "asset_code",
                        "condition" : "in",
                        "value" : ["sinusoid", "Random"]
        },
        "aggregate" : {
                "operation" : "all"
        },
        "timebucket" :  {
                   "timestamp" : "ts",
                   "size"      : "2",
                   "format": "YYYY-MM-DD HH24:MI:SS",
                   "alias": "time"
        },
        "limit" : 10
}
```

Postgres changes and new unit tests will come in next commits

### CODE REVIEW

**Comment by pintomax** [comment]
Example output:

```
{
  "count": 4,
  "rows": [
    {
      "asset_code": "Random",
      "time": "2019-10-24 10:40:44",
      "reading": {
        "random": {
          "min": 100,
          "max": 105,
          "average": 102.2
        }
      }
    },
    {
      "asset_code": "sinusoid",
      "time": "2019-10-24 10:40:40",
      "reading": {
        "ema_1": {
          "min": 0.0167632365,
          "max": 0.2469942184,
          "average": 0.13335103406
        },
        "sinusoid": {
          "min": 0.809016994,
          "max": 0.978147601,
          "average": 0.9035583946
        }
      }
    },
    {
      "asset_code": "Random",
      "time": "2019-10-24 10:40:38",
      "reading": {
        "random": {
          "min": 93,
          "max": 99,
          "average": 95.4
        }
      }
    },
    {
      "asset_code": "Random",
      "time": "2019-10-24 10:40:34",
      "reading": {
        "random": {
          "min": 93,
          "max": 100,
          "average": 97.8
        }
      }
    }
  ]
}
```

File: C/plugins/storage/postgres/connection.cpp [hunk 1]
```diff
@@ -313,6 +531,13 @@ bool Connection::retrieveReadings(const string& condition, string& resultSet)
 				raiseError("retrieve", "Failed to parse JSON payload");
 				return false;
 			}
+
+			// FIXME
```
Comment (MarkRiddoch) [inline C/plugins/storage/postgres/connection.cpp]: What needs to be fixed here?

File: C/plugins/storage/postgres/connection.cpp [hunk 2]
```diff
@@ -505,8 +730,28 @@ bool Connection::retrieveReadings(const string& condition, string& resultSet)
 			{
 				return false;
 			}
+
+			// FIXME
+			if (document.HasMember("aggregate") && document["aggregate"].IsObject())
```
Comment (MarkRiddoch) [inline C/plugins/storage/postgres/connection.cpp]: Likewise, what is the fix?

File: C/plugins/storage/postgres/connection.cpp [hunk 3]
```diff
@@ -1595,6 +1840,8 @@ bool Connection::jsonAggregates(const Value& payload,
 				SQLBuffer& jsonConstraint,
 				bool isTableReading)
 {
+	// FIXME
```
Comment (MarkRiddoch) [inline C/plugins/storage/postgres/connection.cpp]: And the fix is?

File: C/plugins/storage/postgres/connection.cpp [hunk 4]
```diff
@@ -1610,9 +1857,28 @@ bool Connection::jsonAggregates(const Value& payload,
 
 		string column_name = aggregates["column"].GetString();
 
-		sql.append(aggregates["operation"].GetString());
-		sql.append('(');
-		if (aggregates.HasMember("column"))
+		// FIXME
```
Comment (MarkRiddoch) [inline C/plugins/storage/postgres/connection.cpp]: As above, what is needed?

File: C/plugins/storage/postgres/connection.cpp [hunk 5]
```diff
@@ -1610,9 +1857,28 @@ bool Connection::jsonAggregates(const Value& payload,
...
+		if (strcmp(aggregates["operation"].GetString(), "all") == 0)
+		{
+			allOperations = true;
+		}
+
+		// FIXME
```
Comment (MarkRiddoch) [inline C/plugins/storage/postgres/connection.cpp]: As above, what is needed?

File: C/plugins/storage/postgres/connection.cpp [hunk 6]
```diff
@@ -1706,18 +1972,29 @@ bool Connection::jsonAggregates(const Value& payload,
 			}
 			sql.append(")::float");
 		}
-		sql.append(") AS \"");
-		if (aggregates.HasMember("alias"))
+		// FIXME
```
Comment (MarkRiddoch) [inline C/plugins/storage/postgres/connection.cpp]: Same thing?

**Comment by pintomax** [comment]
Sqlite and Postgres tests added

File: C/plugins/storage/postgres/connection.cpp [hunk 7]
```diff
@@ -40,6 +40,222 @@ const vector<string>  pg_column_reserved_words = {
...
+			if (!size)
+			{
+				size = 1;
+			}
+		}
+
+		// Time format for output
+		if (bucket.HasMember("format"))
+		{
+			sql.append("to_char(");
```
Comment (MarkRiddoch) [inline C/plugins/storage/postgres/connection.cpp:115]: Reducing down these calls will help with performance.

**Review by MarkRiddoch — APPROVED** [review]
Looks pretty good to me

File: C/plugins/storage/postgres/connection.cpp (hunk 1 above)
Comment (pintomax) [reply C/plugins/storage/postgres/connection.cpp]: Removed in lastest commit

File: C/plugins/storage/postgres/connection.cpp (hunk 2 above)
Comment (pintomax) [reply C/plugins/storage/postgres/connection.cpp]: Removed in lastest commit

File: C/plugins/storage/postgres/connection.cpp (hunk 4 above)
Comment (pintomax) [reply C/plugins/storage/postgres/connection.cpp]: Removed in lastest commit

File: C/plugins/storage/postgres/connection.cpp (hunk 6 above)
Comment (pintomax) [reply C/plugins/storage/postgres/connection.cpp]: Removed in lastest commit

**Pull request merged**